Dates as Python objects

ntarocco opened the pull request:
* date and datetime fields are now converted to Python datetime objects
  before executing transitions and converted back to ISO format when
  storing the loan record
* add request_start_date to be able to track the start and in the
  end of a PENDING loan (request)
* moves transitions decorators checks from `before` method to
 `execute` to ensure decorators are executed as first thing
* makes configurable to automatically assign the item when creating a
  request via `assign_item` param in the transitions
* adds marshmallow loader for REST endpoint
* adds support to ES7
* changes permissions to take advantage of invenio-access
* ES mappings are now matching schema fields
* Drops support for Python 2
* closes #111, #106, #76, #81


To reviewers: I apologize for the big PR, but I had to change many things after the introduction of the REST loader with fields validation.
Bonus are ES 7 support plus Python 2 drop.

ntarocco commented:
@BadrAly @jma please add anyone else in this PR.

This is the PR to handle dates. I am still fixing tests.
I an a few words:
* there is no easy way to convert dates from string to Python in the marshmallow loader and then simply convert back to string when writing to db via SQLAlchemy. We explored solutions with Invenio Architects but it is not straightforward and we might add support to Invenio in a near future.
* I decided to convert dates just before starting a loan transition, so that we can do date operations (sum/diff/timedelta/comparisons) and then convert back to string in UTC ISO format at the end of the transition just before committing to DB.
* I have updated all tests and added a marshmallow loaded to validate input dates but deserialize them as string
* We have decide to add `request_start_date`: the use case is very similar to the `start_date` of a loan, to avoid to rely on the `created` that can be not aligned with the real creation of the record (e.g. loans request in the past)
* affected fields: as dates only, `start_date`, `end_date`, `request_start_date` and `request_expire_date`, as datetime `transaction_date`
* I have removed `transaction_date` as required param and set it to `utcnow` if not provided.
* I am testing it with our ILS.

kpsherva commented on invenio_circulation/api.py:
```diff
@@ -39,8 +43,24 @@ def create(cls, data, id_=None, **kwargs):
+
         return super(Loan, cls).create(data, id_=id_, **kwargs)
 
+    def dates2datetime(self):
```
I would name it `str2datetime` after reading the docstring

kpsherva commented on invenio_circulation/api.py:
```diff
@@ -39,8 +43,24 @@ def create(cls, data, id_=None, **kwargs):
+            if field in self:
+                self[field] = str2datetime(self[field])
+
+    def dates2str(self):
```
and this probably `datetime2str`

kpsherva commented on invenio_circulation/transitions/base.py:
```diff
@@ -80,6 +79,22 @@ def inner(self, loan, **kwargs):
 
 
+def dates2datetime(f):
+    """Convert loan datetime fields from string to datetime Python objects."""
```
similar here for naming

ntarocco committed:
global: date fields fixes, ES7, drop Python 2

* date and datetime fields are now converted to Python datetime objects
  before executing transitions and converted back to ISO format when
  storing the loan record
* add request_start_date to be able to track the start and in the
  end of a PENDING loan (request)
* moves transitions decorators checks from `before` method to
 `execute` to ensure decorators are executed as first thing
* makes configurable to automatically assign the item when creating a
  request via `assign_item` param in the transitions
* adds marshmallow loader for REST endpoint
* adds support to ES7
* changes permissions to take advantage of invenio-access
* ES mappings are now matching schema fields
* Drops support for Python 2
* closes #111, #106, #76, #81

ntarocco commented on invenio_circulation/api.py:
```diff
@@ -152,29 +194,3 @@ def get_loan_for_item(item_pid):
     return loan
-
-
-def patron_has_active_loan_on_item(patron_pid,
```
**WARNING - Breaking change**
Method dropped because it was wrong: it was checking for `PENDING` and active states, causing lots of issues when using it.

ntarocco commented on invenio_circulation/config.py:
```diff
@@ -32,6 +32,9 @@
 CIRCULATION_DOCUMENT_RETRIEVER_FROM_ITEM = None
 """Function that returns the Document PID of a given Item PID."""
 
+CIRCULATION_STATES_LOAN_REQUEST = ['PENDING']
```
_Reviewer hint_
Introduced variables for the list of states when loan is considered PENDING and CANCELLED so that the system is more configurable

kpsherva replied:
👍

ntarocco commented on invenio_circulation/records/loaders/schemas/json.py:
```diff
@@ -0,0 +1,80 @@
+    def validate_timezone(self, value):
+        """Validate that the passed timezone, if any, is UTC."""
+        # strong validation of input datetime to require UTC timezone
+        if arrow.get(value).isoformat() != \
```
_Reviewer hint_
Force to accept only UTC timezone to minimize timezone issues

ntarocco commented on invenio_circulation/transitions/transitions.py:
```diff
@@ -142,86 +134,77 @@ class ToItemOnLoan(Transition):
+                 **kwargs):
+        super().__init__(src, dest, trigger=trigger,
+                         permission_factory=permission_factory, **kwargs)
+        self.assign_item = kwargs.get('assign_item', True)
```
_Reviewer hint_
Automatic assignment of an item on request is now configurable with this `kwarg`

ntarocco commented on invenio_circulation/transitions/transitions.py:
```diff
@@ -317,14 +296,25 @@ class ItemInTransitHouseToItemReturned(Transition):
+            _update_document_pending_request_for_item(loan['item_pid'])
+
+
+class ToCancelled(Transition):
```
_Reviewer hint_
New transition added: without this, user could by mistake cancel the loan and override the item of the loan
